toc/rfc/rfc-draft-v2-app-manifests.md

```diff
@@ -0,0 +1,613 @@
+      name: my-service-offering
+    service_plan:
+      name: my-service-plan
+    tags:
+      - my-tag
+    parameters:
+      key1: value1
+    metadata:
+      annotations:
+        carrot: service instance metadata
+  - name: user-provided-service-instance
+    type: user-provided
+    credentials:
+      my-cred: secret1234
+    syslog_drain:
+      url: example.com/syslog
+    route_service:
+      url: example.com/route
+  - name: route-service-instance
+    service_offering:
+      name: route-service-offering
+    service_plan:
+      name: route-service-plan
+service_bindings:
+  - type: app
+    service_instance:
+      name: managed-service-instance
+    app:
+      name: app1
+    name: my-service-binding-name
+    parameters:
+      key1: value1
+      key2: value2
+    metadata:
+      annotations:
+        yam: service binding metadata
+  - type: app
+    service_instance:
+      name: managed-service-instance
+    app:
+      name: app2
+  - type: key
+    service_instance:
+      name: managed-service-instance
+    name: my-service-key
+  - type: route
+    service_instance:
+      name: route-service-instance
+    route:
+      url: route.example.com/path
+apps:
+  - name: app1
+    package:
+      path: /local/path/to/source/code
+      metadata:
+        annotations:
+          cassava: package metadata
+    lifecycle:
+      type: buildpack
+      buildpacks:
+        - name: ruby_buildpack
+        - name: java_buildpack
+        - url: git.example.com/my_buildpack
+      stack:
+        name: cflinuxfs4
+    environment_variables:
+      VAR1: value1
+      VAR2: value2
+    features:
+      ssh: true
+    metadata:
+      annotations:
+        contact: bob@example.com jane@example.com
+      labels:
+        sensitive: true
+    processes:
+      - type: web
+        command: start-web.sh
+        scale:
+          disk: 512M
+          instances: 3
+          memory: 500M
+        health_checks:
+          liveness:
+            http_endpoint: /health_check
+            type: http
+            invocation_timeout: 10
+            timeout: 10
+          readiness:
+            type: port
+        metadata:
+          annotations:
+            taro: process metadata
+      - type: worker
+        command: start-worker.sh
+        scale:
+          disk: 1G
+          instances: 2
+          memory: 256M
+        health_checks:
+          liveness:
+            type: process
+            timeout: 15
+  - name: app2
+    state: STOPPED
+    lifecycle:
+      type: docker
+    package:
+      docker:
+        image: docker-image-repository/docker-image-name
+        username: docker-user-name
+    processes:
+      - type: web
+        scale:
+          instances: 1
+          memory: 256M
+    sidecars:
+      - name: authenticator
+        processes:
+          - type: web
+          - type: worker
+        command: bundle exec run-authenticator
+        scale:
+          memory: 800M
+```
+
+#### Notable Changes
```

Comment:
Have you thought to make [Deployments](https://v3-apidocs.cloudfoundry.org/version/release-candidate/#deployments) and [Sidecars](https://v3-apidocs.cloudfoundry.org/version/release-candidate/#sidecars) top level resources? Currently it is not possible with manifest `v1` to define a deployment strategy via the app manifest but this is taken care by the CF CLI. I think it makes sense to declare the deployment strategy explicitly in the app manifest.  Sidecars on the other hand are support by manifest `v1` [sidecar](https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#sidecars) attribute but following the Guiding Principles of this RFC they should be also added as top-level.

Reply:
> I think it makes sense to declare the deployment strategy explicitly in the app manifest.
I think we should support this, but I think we should support setting the default in the CF APIs first. I assume it would be a `default_deployment_strategy` field added to the `App`? IMO we should do that, but drive it end to end through a feature around supporting default deployment strategy generally and add it to both manifest versions.

> Sidecars on the other hand are support by manifest v1 [sidecar](https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#sidecars) attribute but following the Guiding Principles of this RFC they should be also added as top-level.
Sidecars aren't a space-scoped resource, though. They are app-level (highly dependent on what buildpack was used to build the app) and assigned to processes. They're really just glorified additional start commands and resource requests for a process.

Reply:
>Sidecars aren't a space-scoped resource, though. They are app-level (highly dependent on what buildpack was used to build the app) and assigned to processes. They're really just glorified additional start commands and resource requests for a process.

This I understand but my experience is that usually sidecars are used for use cases which apply to all applications in a space like app metrics export setup or app ingress connection management where the sidecar definition is the same. Additionally, sidecars are CF API v3 top-level resource that is why I mentioned it but it is also fine not having a separate definition.

Reply:
Transient resources like deployments and tasks don't fit cleanly into app manifests. 

Perhaps the deployment strategy could be encoded in the `meta` section I proposed in https://github.com/cloudfoundry/community/pull/1019#discussion_r1857298459?